### PR TITLE
Embed endpoints in operations instead of bespoke URL template processing

### LINKED ISSUE #659: DNSSec on Domains Doesn't Work
### What happened?

Tried to enable DNSSec on a Cloud DNS domain

### Steps to reproduce

```
  const zone = new google.dns.v1.ManagedZone(domain.name, {
    dnsName: `${domain.domain}.`,
    dnssecConfig: {
      state: ManagedZoneDnsSecConfigState.Transfer,
    },
    description: "Managed DNS by Pulumi",
  });
```

### Expected Behavior

DNSSec of "Transfer" is visible on the Google Cloud Dashboard, but Pulumi has exited with a failure.

### Actual Behavior

```shell
  google-native:dns/v1:ManagedZone (example):
    error: waiting for completion: polling operation status: googleapi: Error 400: Invalid resource field value in the request.
    Details:
    [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "domain": "googleapis.com",
        "metadata": {
          "method": "cloud.dns.api.v1.ManagedZoneOperationsService.List",
          "service": "dns.googleapis.com"
        },
        "reason": "RESOURCE_PROJECT_INVALID"
      }
    ]
    , invalidParameter
```

### Output of `pulumi about`

_No response_

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by viveklak**
Fixes #659

DNS managed zones and a few other resources need additional information for resolving the operations URL for updates etc than is readily available from the PATCH/Update response. Switching to an embedded endpoint gives us more control over resolving parameters where we can inject parameter values from the resource state or inputs.
The order of resolving URL params is:
1. Response payload from the relevant resource CRUD operation
2. Resource inputs
3. Resource outputs

### CODE REVIEW

File: provider/pkg/gen/overrides.go [hunk 1]
```diff
@@ -51,71 +51,75 @@ var resourceNameByTypeOverrides = map[string]string{
 // that is present in the discovery document.
 var resourceNameByPathOverrides = map[string]string{
 	// Cloud Run: remove domain mapping on the KNative path.
-	"apis/domains.cloudrun.com/v1/namespaces/{namespacesId}/domainmappings": "",
-	"apis/serving.knative.dev/v1/namespaces/{namespacesId}/services":        "",
+	"run_v1.json:apis/domains.cloudrun.com/v1/namespaces/{namespacesId}/domainmappings": "",
```
Comment (viveklak) [inline provider/pkg/gen/overrides.go:54]: These paths were assumed to be globally unique which is not the case.

File: provider/pkg/gen/discovery.go [hunk 2]
```diff
@@ -142,6 +155,19 @@ func findResourcesImpl(docName, parentName string, rest map[string]discovery.Res
...
+				operationGetPath := getMethod.FlatPath
+				if operationGetPath == "" {
+					operationGetPath = getMethod.Path
+				}
+				if override, has := resourceNameByPathOverrides[fmt.Sprintf("%s:%s", filepath.Base(docName),
+					operationGetPath)]; has {
+					if override == "" {
+						continue
+					}
+					name = override
```
Comment (lblackstone) [inline provider/pkg/gen/discovery.go:168]:
```suggestion
					if override != "" {
						name = override
					}
```

File: provider/pkg/gen/schema_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
+package gen
```
Comment (lblackstone) [inline provider/pkg/gen/schema_test.go:1]: Missing copyright

File: provider/pkg/gen/schema_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+				paramNames := codegen.NewStringSet(tmpl.Names()...)
+
+				var referencedOp *operation
+				switch resourceOperationType {
+				case "CREATE":
+					if ddResource.createMethod.Response.Ref != "" {
+						if createOp, ok := ops[ddResource.createMethod.Response.Ref]; ok {
+							referencedOp = createOp
+						}
+					}
```
Comment (lblackstone) [inline provider/pkg/gen/schema_test.go:89]: Do we also need to validate READ?

File: provider/pkg/gen/schema_test.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+					if found {
+						continue
+					}
+
+					for name := range resource.Properties {
+						if name == expectedName {
+							found = true
+							break
+						}
+					}
```
Comment (lblackstone) [inline provider/pkg/gen/schema_test.go:143]: Is the `if found` skip supposed to appear before the `resource.Properties` check?

File: provider/pkg/gen/schema_test.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		}
+
+	}
+}
+
+// loadMetadata deserializes the provided compressed json byte array into a CloudAPIMetadata struct.
+func loadMetadata() (*resources.CloudAPIMetadata, error) {
+	var resourceMap resources.CloudAPIMetadata
+
+	bytes, err := os.Open(path.Join(root, "provider", "cmd", "pulumi-resource-google-native",
```
Comment (lblackstone) [inline provider/pkg/gen/schema_test.go:166]:
```suggestion
	bytes, err := os.Open(filepath.Join(root, "provider", "cmd", "pulumi-resource-google-native",
```

File: provider/pkg/gen/schema_test.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	if err = json.NewDecoder(bytes).Decode(&resourceMap); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling resource map")
+	}
+	return &resourceMap, nil
+}
+
+func loadSchema() (*schema.PackageSpec, error) {
+	var pkg schema.PackageSpec
+
+	bytes, err := os.Open(path.Join(root, "provider", "cmd", "pulumi-resource-google-native",
```
Comment (lblackstone) [inline provider/pkg/gen/schema_test.go:180]:
```suggestion
	bytes, err := os.Open(filepath.Join(root, "provider", "cmd", "pulumi-resource-google-native",
```

File: provider/pkg/provider/provider.go [hunk 8]
```diff
@@ -916,19 +916,21 @@ func (p *googleCloudProvider) waitForResourceOpCompletion(
 				}
 				// At this point, we assume either a complete failure or a clean response.
 				if !continuePollingForRestingState {
+					// Try and fetch state from
```
Comment (lblackstone) [inline provider/pkg/provider/provider.go:919]: Looks like the comment was cut off

**Review by lblackstone — CHANGES_REQUESTED** [review]

File: provider/pkg/gen/schema_test.go (hunk 4 above)
Comment (viveklak) [reply provider/pkg/gen/schema_test.go:89]: Reads don't result in delayed/long running operations like the other verbs.

File: provider/pkg/provider/provider.go (hunk 8 above)
Comment (viveklak) [reply provider/pkg/provider/provider.go]: Fixed.

File: provider/pkg/gen/schema_test.go (hunk 6 above)
Comment (viveklak) [reply provider/pkg/gen/schema_test.go]: Fixed. Thanks for pointing out.

File: provider/pkg/gen/schema_test.go (hunk 5 above)
Comment (viveklak) [reply provider/pkg/gen/schema_test.go]: Yeah. I have added a few more comments to the test to make the intent here clearer. Could you take another look?

File: provider/pkg/gen/schema_test.go (hunk 3 above)
Comment (viveklak) [reply provider/pkg/gen/schema_test.go:15]: Fixed.

File: provider/pkg/gen/discovery.go (hunk 2 above)
Comment (viveklak) [reply provider/pkg/gen/discovery.go:170]: I have added a few more comments here which hopefully help. The intent with an empty string value is to ignore the specified resource/operation.

**Review by lblackstone — APPROVED** [review]

**Pull request merged**